File: lib/fsharp/Common_contract.py
from __future__ import annotations
from collections.abc import Callable
from typing import Any
from fable_modules.fable_library.reflection import (TypeInfo, union_type)
from fable_modules.fable_library.types import (Array, Union)
from ......lib.spiral.lib import SpiralTrace_trace
from ......lib.spiral.trace import US0

nl: str = "\n"

q: str = "\""

def _expr156() -> TypeInfo:
    return union_type("Polyglot.Common.TraceLevel", [], TraceLevel, lambda: [[], [], [], [], []])


class TraceLevel(Union):
    def __init__(self, tag: int, *fields: Any) -> None:
        super().__init__()
        self.tag: int = tag or 0
        self.fields: Array[Any] = list(fields)

    @staticmethod
    def cases() -> list[str]:
        return ["Verbose", "Debug", "Info", "Warning", "Critical"]


TraceLevel_reflection = _expr156

def TraceLevel__get_IsVerbose(this: TraceLevel, unit_arg: None) -> bool:
    if this.tag == 0:
        return True

    else: 
        return False



def TraceLevel__get_IsDebug(this: TraceLevel, unit_arg: None) -> bool:
    if this.tag == 1:
        return True

    else: 
        return False



def TraceLevel__get_IsInfo(this: TraceLevel, unit_arg: None) -> bool:
    if this.tag == 2:
        return True

    else: 
        return False



def TraceLevel__get_IsWarning(this: TraceLevel, unit_arg: None) -> bool:
    if this.tag == 3:
        return True

    else: 
        return False



def TraceLevel__get_IsCritical(this: TraceLevel, unit_arg: None) -> bool:
    if this.tag == 4:
        return True

    else: 
        return False



def to_trace_level(_arg: TraceLevel) -> US0:
    if _arg.tag == 1:
        return US0(1)

    elif _arg.tag == 2:
        return US0(2)

    elif _arg.tag == 3:
        return US0(3)

    elif _arg.tag == 4:
        return US0(4)

    else: 
        return US0(0)



def trace(level: TraceLevel, fn: Callable[[], str], locals: Callable[[], str]) -> None:
    SpiralTrace_trace(to_trace_level(level))(fn)(locals)

File: lib/fsharp/Common_contract.dart
// ignore_for_file: camel_case_types, constant_identifier_names, non_constant_identifier_names, unnecessary_this
import '../../fable_modules/fable_library/Types.dart' as types;
import '../../../../../../../../lib/spiral/lib.fsx' as lib;
import '../../../../../../../../lib/spiral/trace.fsx' as trace_1;

const nl = '\n';

const q = '\"';

class TraceLevel implements types.Union, Comparable<TraceLevel> {
    final int tag;
    const TraceLevel(this.tag);
    @override
    bool operator ==(Object other) => (other is TraceLevel) && (other.tag == tag);
    @override
    int get hashCode => tag.hashCode;
    @override
    int compareTo(TraceLevel other) => tag.compareTo(other.tag);
}

bool TraceLevel__get_IsVerbose(TraceLevel this$, void unitArg) {
    if (this$.tag == /* Verbose */ 0) {
        return true;
    } else {
        return false;
    }
}

bool TraceLevel__get_IsDebug(TraceLevel this$, void unitArg) {
    if (this$.tag == /* Debug */ 1) {
        return true;
    } else {
        return false;
    }
}

bool TraceLevel__get_IsInfo(TraceLevel this$, void unitArg) {
    if (this$.tag == /* Info */ 2) {
        return true;
    } else {
        return false;
    }
}

bool TraceLevel__get_IsWarning(TraceLevel this$, void unitArg) {
    if (this$.tag == /* Warning */ 3) {
        return true;
    } else {
        return false;
    }
}

bool TraceLevel__get_IsCritical(TraceLevel this$, void unitArg) {
    if (this$.tag == /* Critical */ 4) {
        return true;
    } else {
        return false;
    }
}

trace_1.US0 to_trace_level(TraceLevel _arg) {
    switch (_arg.tag) {
        case 1:
            return const trace_1.US0(/* US0_1 */ 1);
        case 2:
            return const trace_1.US0(/* US0_2 */ 2);
        case 3:
            return const trace_1.US0(/* US0_3 */ 3);
        case 4:
            return const trace_1.US0(/* US0_4 */ 4);
        default:
            return const trace_1.US0(/* US0_0 */ 0);
    }
}

void trace(TraceLevel level, String Function() fn, String Function() getLocals) {
    ((lib.SpiralTrace_trace(to_trace_level(level)))(fn))(getLocals);
}

